File: app/src/main/java/com/savory/login/LoginClient.java
package com.savory.login;

import android.app.Activity;
import android.content.Intent;
import android.support.annotation.NonNull;

import com.facebook.FacebookCallback;
import com.facebook.FacebookException;
import com.facebook.login.LoginResult;
import com.savory.api.clients.savory.SavoryClient;
import com.savory.api.clients.savory.models.SavoryToken;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class LoginClient {

    interface LoginListener {
        void onLoginPending();

        void onSuccessfulLogin(SavoryToken savoryToken);

        void onLoginCancelled();

        void onLoginError();
    }

    protected SavoryClient savoryClient;
    protected Call<SavoryToken> savoryTokenCall;
    protected LoginListener listener;

    private FacebookClient facebookClient = new FacebookClient();
    private final FacebookCallback<LoginResult> fbLoginCallback = new FacebookCallback<LoginResult>() {
        @Override
        public void onSuccess(LoginResult loginResult) {
            if (listener != null) {
                listener.onLoginPending();
                savoryTokenCall = savoryClient.connect(
                    loginResult.getAccessToken().getToken());
                savoryTokenCall.enqueue(new Callback<SavoryToken>() {
                    @Override
                    public void onResponse(Call<SavoryToken> call, Response<SavoryToken> response) {
                        if (listener != null) {
                            if (response.isSuccessful()) {
                                listener.onSuccessfulLogin(response.body());
                            } else {
                                listener.onLoginError();
                            }
                        }
                    }

                    @Override
                    public void onFailure(Call<SavoryToken> call, Throwable t) {
                        if (listener != null) {
                            listener.onLoginError();
                        }
                    }
                });
            }
        }

        @Override
        public void onCancel() {
            if (listener != null) {
                listener.onLoginCancelled();
            }
        }

        @Override
        public void onError(FacebookException error) {
            if (listener != null) {
                listener.onLoginError();
            }
        }
    };

    public LoginClient(@NonNull SavoryClient savoryClient) {
        this.savoryClient = savoryClient;
        facebookClient.registerCallback(fbLoginCallback);
    }

    void loginWithFacebook(Activity activity) {
        facebookClient.login(activity);
    }

    void setListener(@NonNull LoginListener listener) {
        this.listener = listener;
    }

    void onLoginResult(int requestCode, int resultCode, Intent data) {
        facebookClient.onLoginResult(requestCode, resultCode, data);
    }
}

File: app/src/main/java/com/savory/login/LoginClient.kt
package com.savory.login

import android.content.Intent
import com.facebook.FacebookCallback
import com.facebook.FacebookException
import com.facebook.login.LoginResult
import com.savory.api.SavoryClient
import com.savory.api.models.SavoryToken
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.lang.Exception

class LoginClient(val savoryClient: SavoryClient) {

    interface LoginListener {
        fun onSuccessfulLogin(savoryToken : SavoryToken)
        fun onLoginCancelled()
        fun onLoginError(throwable: Throwable)
    }

    val facebookClient = FacebookClient()
    var listener : LoginListener? = null
    var savoryTokenCall : Call<SavoryToken>? = null

    val facebookLoginCallback = object : FacebookCallback<LoginResult> {
        override fun onSuccess(result: LoginResult?) {
            val localSavoryTokenCall = savoryClient.connect(result!!.accessToken.token.toString());
            savoryTokenCall = localSavoryTokenCall
            localSavoryTokenCall.enqueue(object : Callback<SavoryToken> {
                override fun onResponse(call: Call<SavoryToken>,
                                        response: Response<SavoryToken>) {
                    if (response.isSuccessful()) {
                        listener?.onSuccessfulLogin(response.body()!!)
                    } else {
                        listener?.onLoginError(Exception("server error"))
                    }
                }

                override fun onFailure(call: Call<SavoryToken>, t: Throwable) {
                    listener?.onLoginError(t)
                }
            })
        }

        override fun onCancel() {
            listener?.onLoginCancelled()
        }

        override fun onError(error: FacebookException) {
            listener?.onLoginError(error)
        }
    }

    init {
        facebookClient.registerCallback(facebookLoginCallback)
    }

    fun bindFacebookButton(facebookButton: FacebookButton) {
        facebookClient.bind(facebookButton)
    }

    fun listen(listener: LoginListener) {
        this.listener = listener
    }

    fun onLoginResult(requestCode: Int, resultCode: Int, data: Intent) {
        facebookClient.onLoginResult(requestCode, resultCode, data)
    }
}